### PR TITLE
Implement waiting room

### PR DESCRIPTION
**Pull request by bpaquet**
Implement the waiting room api in order to add them in the terraform cloudflare provider

### CODE REVIEW

**Comment by bpaquet** [comment]
Test are missing, I can add them if this is the only missing point :)

**Comment by bpaquet** [comment]
Tests are ready @jacobbednarz :)

File: waiting_room.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	Errors   []string      `json:"errors"`
+	Messages []string      `json:"messages"`
+	Result   []WaitingRoom `json:"result"`
+}
+
+// CreateWaitingRoom creates a new Waiting Room for a zone.
+//
+// API reference: https://api.cloudflare.com/#waiting-room-create-waiting-room
+func (api *API) CreateWaitingRoom(ctx context.Context, zoneID string, waitingRoom WaitingRoom) (*WaitingRoom, error) {
+	uri := "/zones/" + zoneID + "/waiting_rooms"
```
Comment (jacobbednarz) [inline waiting_room.go:50]: #626 consolidated URLs to use `fmt.Sprintf` so we'll update that here too

```suggestion
	uri := fmt.Sprintf("/zones/%s/waiting_rooms", zoneID)
```

File: waiting_room.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		return nil, errors.Wrap(err, errUnmarshalError)
+	}
+	return &r.Result, nil
+}
+
+// ListWaitingRooms returns all Waiting Room for a zone.
+//
+// API reference: https://api.cloudflare.com/#waiting-room-list-waiting-rooms
+func (api *API) ListWaitingRooms(ctx context.Context, zoneID string) ([]WaitingRoom, error) {
+	uri := "/zones/" + zoneID + "/waiting_rooms"
```
Comment (jacobbednarz) [inline waiting_room.go:67]:
```suggestion
	uri := fmt.Sprinft("/zones/%s/waiting_rooms", zoneID)
```

File: waiting_room.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	return r.Result, nil
+}
+
+// UpdateWaitingRoom lets you replace a Waiting Room. This is in contrast to
+// ChangeWaitingRoom which lets you change individual settings.
+//
+// API reference: https://api.cloudflare.com/#waiting-room-update-waiting-room
+func (api *API) UpdateWaitingRoom(ctx context.Context, zoneID string, waitingRoom WaitingRoom) (WaitingRoom, error) {
+	uri := "/zones/" + zoneID + "/waiting_rooms/" + waitingRoom.ID
```
Comment (jacobbednarz) [inline waiting_room.go:120]:
```suggestion
	uri := fmt.Sprinft("/zones/%s/waiting_rooms/%s", zoneID, waitingRoom.ID)
```

File: waiting_room.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		return WaitingRoom{}, errors.Wrap(err, errUnmarshalError)
+	}
+	return r.Result, nil
+}
+
+// DeleteWaitingRoom deletes a Waiting Room for a zone.
+//
+// API reference: https://api.cloudflare.com/#waiting-room-delete-waiting-room
+func (api *API) DeleteWaitingRoom(ctx context.Context, zoneID, waitingRoomId string) error {
+	uri := "/zones/" + zoneID + "/waiting_rooms/" + waitingRoomId
```
Comment (jacobbednarz) [inline waiting_room.go:137]:
```suggestion
func (api *API) DeleteWaitingRoom(ctx context.Context, zoneID, waitingRoomID string) error {
	uri := fmt.Sprintf("/zones/%s/waiting_rooms/%s", zoneID, waitingRoomID)
```

File: waiting_room.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	return r.Result, nil
+}
+
+// ChangeWaitingRoom lets you change individual settings for a Waiting room. This is
+// in contrast to UpdateWaitingRoom which replaces the entire Waiting room.
+//
+// API reference: https://api.cloudflare.com/#waiting-room-update-waiting-room
+func (api *API) ChangeWaitingRoom(ctx context.Context, zoneID, waitingRoomId string, waitingRoom WaitingRoom) (WaitingRoom, error) {
+	uri := "/zones/" + zoneID + "/waiting_rooms/" + waitingRoomId
```
Comment (jacobbednarz) [inline waiting_room.go:102]:
```suggestion
func (api *API) ChangeWaitingRoom(ctx context.Context, zoneID, waitingRoomID string, waitingRoom WaitingRoom) (WaitingRoom, error) {
	uri := "/zones/%s/waiting_rooms/%s", zoneID, waitingRoomID)
```

File: waiting_room.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		return []WaitingRoom{}, errors.Wrap(err, errUnmarshalError)
+	}
+	return r.Result, nil
+}
+
+// WaitingRoom fetches detail about one Waiting room for a zone.
+//
+// API reference: https://api.cloudflare.com/#waiting-room-waiting-room-details
+func (api *API) WaitingRoom(ctx context.Context, zoneID, waitingRoomId string) (WaitingRoom, error) {
+	uri := "/zones/" + zoneID + "/waiting_rooms/" + waitingRoomId
```
Comment (jacobbednarz) [inline waiting_room.go:84]:
```suggestion
func (api *API) WaitingRoom(ctx context.Context, zoneID, waitingRoomID string) (WaitingRoom, error) {
	uri := fmt.Sprintf("/zones/%s/waiting_rooms/%s, zoneID, waitingRoomID)
```

File: waiting_room.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	SessionDuration       int       `json:"session_duration"`
+	DisableSessionRenewal bool      `json:"disable_session_renewal"`
+	CustomPageHtml        string    `json:"custom_page_html,omitempty"`
+}
+
+// WaitingRoomDetailResponse is the API response, containing a single WaitingRoom.
+type WaitingRoomDetailResponse struct {
+	Success  bool        `json:"success"`
+	Errors   []string    `json:"errors"`
+	Messages []string    `json:"messages"`
```
[the remark this answers is not in the document]
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	Result   WaitingRoom `json:"result"`
+}
+
+// WaitingRoomsResponse is the API response, containing an array of WaitingRooms.
+type WaitingRoomsResponse struct {
+	Success  bool          `json:"success"`
+	Errors   []string      `json:"errors"`
+	Messages []string      `json:"messages"`
```
[the remark this answers is not in the document]
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go (hunk 1 above)
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go (hunk 2 above)
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go (hunk 3 above)
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go (hunk 4 above)
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go (hunk 5 above)
Comment (bpaquet) [reply waiting_room.go]: Fixed

File: waiting_room.go (hunk 6 above)
Comment (bpaquet) [reply waiting_room.go]: Fixed

**Comment by bpaquet** [comment]
@jacobbednarz any way to release a 1.17 version?

**Comment by jacobbednarz** [comment]
as this is a new feature and non urgent, it will go out with the regular release sometime in the next week or so depending on other changes.

**Pull request merged**